Initial Commit: Implement Install Resource

alecxvs opened the pull request:
This is the initial implementation of the resource cookbook, including the install resource for installing and uninstalling Splunk and appropriate tests.

Most potential issues may be in formatting, style, and documentation.

alecxvs commented on test/integration/install/spec.rb:
```diff
@@ -0,0 +1,46 @@
+    it { is_expected.to exist }
+    its('group') { is_expected.to eq 'splunk' } unless windows
+  end
+end
```
Inconsistent newlines

alecxvs replied:
dbadf53922e6dcc4b16f4829a189d31f84255b63

alecxvs commented on test/integration/install/spec.rb:
```diff
@@ -0,0 +1,46 @@
+# Encoding: utf-8
```
Line after Encoding inconsistent in most of the files. There should probably be a new line here.

alecxvs replied:
dbadf53922e6dcc4b16f4829a189d31f84255b63

alecxvs commented on resources/splunk_install.rb:
```diff
@@ -0,0 +1,220 @@
+
+class DebianInstall < SplunkInstall
+  resource_name :splunk_install
+  # use_automatic_resource_name
```
Remove

alecxvs replied:
dbadf53922e6dcc4b16f4829a189d31f84255b63

alecxvs commented on README.md:
```diff
@@ -0,0 +1,68 @@
+
+License and Authors
+-------------------
+- Author:: Alec Sears (alec.sears@cerner.com)
```
Add more people?

acharlieh replied:
You're the only author so far :) we'll add more as more people contribute significantly to it.

acharlieh commented:
Let's break up commits in a more logical manner... like "Basic project structure" and "Install Resource"

acharlieh commented on .kitchen.yml:
```diff
@@ -0,0 +1,33 @@
+  - name: windows
+    driver:
+      box: win2012r2-standard-nocm-1.1.0
+      box_url: http://repo.release.cerner.corp/nexus/content/repositories/vagrant/com/cerner/vagrant/win2012r2-standard-nocm/1.1.0/win2012r2-standard-nocm-1.1.0.box
```
This isn't public, so we can't use it. instead what about using a box of opentable/win-2012r2-standard-amd64-nocm 

https://atlas.hashicorp.com/opentable/boxes/win-2012r2-standard-amd64-nocm/

alecxvs replied:
6a1302f202a615c5d2a4d9fe7f110e9dbca57f3d
bd7246d83652b4d8350e3a64431a7e0874cfe7f1

acharlieh commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,35 @@
+    - The merge commit text should reference the issue and the pull request so that both are closed when the merge is pushed.
+
+# Release Process:
+1. A pull request with the release commit is made to the 'stable' branch.
```
hm... we should update this text in the cerner_splunk one, because we no longer use a PR for releases since we increment version at merge of pull request time so release is just tagging. We use an issue proposing the tag sha, tag text, issue labeled release, and summarize changes.

alecxvs replied:
It sounds like that should be the case here, too:
679ccab285a18216eb8f6aacbcefd4ed4fb69954

acharlieh commented on README.md:
```diff
@@ -0,0 +1,68 @@
+| build            |               String                |       **Yes**        |                                         | Build number of the version                                                                                                                                                                                                                                              |
+| service_user     |               String                |          No          | `'splunk'`                              | User to run the Splunk process                                                                                                                                                                                                                                           |
+| service_password |               String                |          No          |                                         | Shadow hash of the password for the user to run the Splunk process. The resource will attempt to install Splunk as a system user, but some platforms or configurations may require a password.                                                                           |
+| source_url       |               String                |          No          | `'http://download.splunk.com/products'` | Base url to pull Splunk packages from. Use this if you are mirroring the downloads for Splunk packages. The resource will append the version, os, and filename to the url like so: `{source_url}/splunk/releases/0.0.0/linux/splunk-0.0.0-a1b2c3d4e5f6-Linux-x86_64.tgz` |
```
We should default this to the https url.

alecxvs replied:
6a1302f202a615c5d2a4d9fe7f110e9dbca57f3d

acharlieh commented on resources/splunk_install.rb:
```diff
@@ -0,0 +1,221 @@
+    else raise 'Package must be specified (:splunk or :universal_forwarder)'
+    end unless package
+
+    raise "Invalid package: #{package}" unless %i(splunk universal_forwarder).include? package
```
Wouldn't this be covered by the validation of the package property?

alecxvs replied:
4867bb4

alecxvs commented:
Initial commit rebased into multiple:

Log message | Commit Hash
-------------------|-------------------
Initial Project Structure | 9b8c865
Implement splunk_install resource | f9b7eac
Unit Tests | 286a1e5
Integration Tests | cf63718

nmattam commented:
Looks like your email is not linked to the commits.

alecxvs commented:
Fixed commit author

acharlieh commented on metadata.rb:
```diff
@@ -0,0 +1,20 @@
+supports         'ubuntu', '>= 12.04'
+supports         'windows', '>= 6.1'
+
+# depends          'poise-service', '~> 1.3.0' #Will use this for future service resource
```
I would remove it for right now then. No need to have an extra comment here.

alecxvs replied:
f4ec2dbe9b3040fe83d6d2143797c9c9555684ca

acharlieh commented on libraries/path_helpers.rb:
```diff
@@ -0,0 +1,24 @@
+# Encoding: utf-8
+
+module PathHelpers
```
General comment... should we be name-spacing our classes and modules in a module named for the cookbook so that generic names like "PathHelpers" aren't accidentally reused or worse accidentally monkey patched elsewhere?

alecxvs replied:
Yeah, that's probably a good idea. I was thinking of the modules organizationally at first, which that doesn't quite work now that I have two helper modules.

alecxvs replied:
f870316e401a5b4fb7cd81171497d0bebe208615

acharlieh commented on resources/splunk_install.rb:
```diff
@@ -0,0 +1,219 @@
+  end
+
+  # kernel_string = x64_support ? 'windows-64' : 'windows-32'
+  # extension = 'zip'
```
also comments from development that can be cleaned up.

alecxvs replied:
f01e160

acharlieh commented on resources/splunk_install.rb:
```diff
@@ -0,0 +1,219 @@
+  end
+
+  action :install do
+    raise 'Windows installation will fail unless license is accepted.' unless accept_license
```
Do we even want to provide the option to not programmatically accept the license then?

alecxvs replied:
Fair point, it can be made required with a true value only since it's not used elsewhere
That makes me wonder if we should force license agreement before installing on any platform in the future (if that's even possible considering that it has to be done on startup on linux)

alecxvs replied:
c2e8a3bdf6cd33e9779c4cc76d0802e600738ca9

acharlieh commented on resources/splunk_install.rb:
```diff
@@ -0,0 +1,219 @@
+    end
+  end
+
+  action :uninstall do
```
Should you be calling super after invoking remove on the package resource?

alecxvs replied:
If we want to be super sure the splunk directory gets deleted, then yes. Package removal should be deleting the directory (or most of it, I think the rpm removal leaves a single redhat specific script)

alecxvs replied:
I ended up super()ing anyways to share run_state changes
c2e8a3bdf6cd33e9779c4cc76d0802e600738ca9

acharlieh commented on resources/splunk_install.rb:
```diff
@@ -0,0 +1,219 @@
+  property :service_password, String
+  property :version, String, required: true
+  property :build, String, required: true
+  property :source_url, String, default: 'https://download.splunk.com/products'
```
I might be being pedantic, but is this a source url or mirror url or a base url?

alecxvs replied:
It feels like it could be any of the three, but it might be leaning towards a base URL.

alecxvs replied:
aa54e36fa390277513a97ff3bc03d9dcc923ad59

acharlieh commented on resources/splunk_install.rb:
```diff
@@ -0,0 +1,219 @@
+
+  property :name, String, name_property: true, identity: true
+  property :package, [:splunk, :universal_forwarder], required: true
+  property :service_user, String, default: 'splunk'
```
do we want to manage group as well? or no?

alecxvs replied:
Is there a use for doing so? Having a system user assumes there's a group named after the user, so that's not going to change. We could allow adding extra groups, but consumers could also do that. It might avoid cloning the user resource, however.

acharlieh replied:
Hm...good point... Do we do anything with the user in this resource other than ensure it's created? Do we need it for install time on Windows (I don't think I see it being used)? Or could user / service management for an installation be a different resource or different action? (action :enable / :disable to manage the service?)

alecxvs replied:
We don't seem to be using the user at all. Windows doesn't require it, but it needs to be set at install time I think. User creation should be part of service setup (which is intertwined with the Windows installation process but we can deal with that when we get to it).

I'm probably going to remove user creation then and it will be part of the service resource.

alecxvs replied:
c2e8a3bdf6cd33e9779c4cc76d0802e600738ca9

acharlieh commented on test/integration/install/spec.rb:
```diff
@@ -0,0 +1,51 @@
+  title 'Install Splunk'
+  tag 'splunk_install'
+
+  describe package(os.windows? ? 'Splunk Enterprise' : 'splunk') do
```
nit... you defined windows as `os.windows?` above.... do you want to use it here? or is it not saving you enough that you should just use os.windows? on lines 21 and 25

alecxvs replied:
`windows` should be used.
Why I'm assigning `windows` instead of just using `os.windows?` isn't to save space, but because `os` isn't defined specifically when I use it here: `describe ... end unless os.windows?`
See: https://github.com/chef/inspec/issues/628

alecxvs replied:
f01e16056a3dbabbd9315055660a06bebd8689e7

acharlieh commented on test/unit/spec/default_spec.rb:
```diff
@@ -0,0 +1,260 @@
+    }
+  }
+}
+describe 'splunk_install' do
```
Nit: probably want a newline between your hash definition and your spec.

alecxvs replied:
f01e16056a3dbabbd9315055660a06bebd8689e7

acharlieh commented on .travis.yml:
```diff
@@ -0,0 +1,12 @@
+cache: bundler
+sudo: false
+rvm:
+- 2.1.6
```
So you have 2.1.6 here... current version of chefdk ships 2.1.8... but your Gemfile doesn't install on either of these: 

    ...
    Using ruby_dep 1.3.1
    Installing listen 3.1.5

    Gem::InstallError: listen requires Ruby version >= 2.2.3, ~> 2.2.
    An error occurred while installing listen (3.1.5), and Bundler cannot continue.
    Make sure that `gem install listen -v '3.1.5'` succeeds before bundling.

alecxvs replied:
f01e16056a3dbabbd9315055660a06bebd8689e7

Should it match ChefDK version?

acharlieh commented on README.md:
```diff
@@ -0,0 +1,68 @@
+---------
+
+### splunk_install
+Installs Splunk. Easy.
```
"Use the splunk_install resource to manage the core parts of a Splunk installation" perhaps? The resource itself doesn't install splunk, using one of the actions you take on the resource asserts that the desired version of Splunk is installed (subject to package manager constraints of course). (Also I'm not sure if I'm crazy or not with this idea, but I wonder if like with ReST, Resources should be Nouns, and you perform Action (verbs) upon them.)

alecxvs replied:
I wanted to go the route of nouns and verbs, but opted for splunk_install as it was probably going to be a huge resource to cram configure in with it. Chances are, though, that the resource-side of things isn't going to be terribly complicated. I fully intend to rename the resource to just 'splunk' if I add configuration and it's clean enough.

alecxvs replied:
c2e8a3bdf6cd33e9779c4cc76d0802e600738ca9

acharlieh commented on metadata.rb:
```diff
@@ -0,0 +1,19 @@
+license          'Apache 2.0'
+description      'Installs Splunk'
+long_description IO.read(File.join(File.dirname(__FILE__), 'README.md'))
+version          '1.0.0'
```
Something to think about, do we want to start at 1.0.0 or do we want to start with 0.1.0 ?

alecxvs replied:
If we're going to be making changes parallel to cerner_splunk, 0.1.0 is a better choice.

alecxvs replied:
f01e16056a3dbabbd9315055660a06bebd8689e7

acharlieh commented on resources/splunk_install.rb:
```diff
@@ -0,0 +1,219 @@
+    state = node.run_state['splunk_ingredient'] ||= {}
+    state['packages'] ||= {}
+
+    raise "Installation resource with name #{name} already exists!" if state['packages'][name]
```
should we be keying by name? or should we key by install_dir ?

acharlieh replied:
We probably should be removing information from run state in the case of an uninstallation as well?

alecxvs replied:
We should key by install_dir. I forgot I moved this check to install, and I wanted to keep the name so I could easily test uninstall.

We should probably store installations in run_state instead of packages, as that's what the splunk_install resource deals with. I'm curious now, though, if I could/should load existing installations into run_state at compile time.

alecxvs replied:
c2e8a3bdf6cd33e9779c4cc76d0802e600738ca9

acharlieh commented on resources/splunk_install.rb:
```diff
@@ -0,0 +1,219 @@
+      package: package,
+      version: version,
+      build: build,
+      x64: x64_support
```
I'm not sure we need this, as this comes simply from an ohai attribute yes?

alecxvs replied:
It's not as clear since it could be represented as x86_64 or amd64 (or other combinations of weirdness on other platforms that aren't supported). Though we don't really need to provide this in the run state, I'm already producing a boolean for 64-bit support and it doesn't hurt to throw that in with the package state (plus it makes it clear we're installing the 64-bit version of Splunk and is more easily testable).

alecxvs commented on .travis.yml:
```diff
@@ -0,0 +1,12 @@
+script:
+- bundle exec rubocop
+- bundle exec foodcritic -f any .
+- bundle exec rspec test/unit
```
Should I add kitchen tests to this?

alecxvs replied:
Also rubocop is being run on all the gems in the bundle, which is pointless.

acharlieh replied:
It'd be nice if we could do kitchen tests as well, we would need to figure out how to do kitchen tests on Travis (or another cloud based CI system) and ensure it's not brittle. Just not something we've done yet, but it'd be cool if we could.

alecxvs replied:
Is there a way to run manual travis checks with different configs? I'd like to play with it a bit to try and get a chef-zero run and run inspec, if not kitchen as a whole (there's no way to run vagrant boxes on travis, so we would end up running chef locally and testing that either way)

alecxvs commented:
I added unit tests for removing run_state on uninstall, and changed how I was stubbing the version file in another test. Instead of explicitly calling the chef runner again, it now does the stubbing in a let that's called before the runner runs. I think it's still running for each individual example, but there's less mess in the code now and less potential for it to be running redundantly.

Also renamed state to install_state (in splunk_install.rb) because apparently state is something defined by the Chef Resource class.

9becf50fd55ab778bc0dbe49574c952a096e11c7

nmattam commented:
+1

acharlieh commented on libraries/path_helpers.rb:
```diff
@@ -0,0 +1,40 @@
+    #
+    # @param splunk_home [String] the absolute path to Splunk's home directory
+    # @return [Pathname] the absolute path to Splunk's version file
+    def version_pathname(splunk_home = install_dir)
```
Where does this default value come from?

alecxvs replied:
I'm not sure how to describe it in the documentation, but it's a little magic because that's a method defined in splunk_install where the resource is being mixed in. There's no reason to be doing that though since I call it maybe once, I can pass the parameter normally.

alecxvs replied:
34547d81c56dd573c7e431b0ddb8d1939a0d07dd

acharlieh commented on resources/splunk_install.rb:
```diff
@@ -0,0 +1,223 @@
+
+  action :uninstall do
+    install_state = node.run_state['splunk_ingredient']
+    install_state.delete('current_installation')
```
Should you be removing this only if this resource corresponds to the current_installation?

alecxvs replied:
That's a curious question, because it's easier and more sanitary in my mind to remove the current_installation as that entry in run_state is meant for chaining future resources and performing actions immediately after an install. It would be very unusual to do so after an uninstall, and I don't think we need that logic right now. It's not an impossible scenario, though, either. Just maybe one that has some smelly code.

Furthermore, I intend to follow the pattern of setting current_installation in every resource that specifies an installation, so with uninstall not having an installation anymore to set this to, it makes more sense that there is no current_installation.

jdmurphy commented on Gemfile:
```diff
@@ -0,0 +1,18 @@
+source 'https://rubygems.org'
+
+gem 'berkshelf', '> 3.1'
```
Minor in the scope of things, but considering the issues we have run into with the 3.x series (Faraday gzip comes to mind), I would stick to the 4.x series of berkshelf.

alecxvs replied:
0551e39585dea006cbcbf284a30e5bf1ddfbb5f9

jdmurphy commented on README.md:
```diff
@@ -0,0 +1,66 @@
+---------
+
+### splunk_install
+Manages an installation of splunk
```
Minor, capital S on Splunk.

alecxvs replied:
0551e39585dea006cbcbf284a30e5bf1ddfbb5f9

jdmurphy commented on resources/splunk_install.rb:
```diff
@@ -0,0 +1,223 @@
+  property :version, String, required: true
+  property :build, String, required: true
+  property :base_url, String, default: 'https://download.splunk.com/products'
+  property :accept_license, [TrueClass], required: true
```
Don't we need `[TrueClass, FalseClass]` here to be able to do either true or false?

alecxvs replied:
We would throw an error when giving false anyways, might as well fail fast by failing the validation.

jdmurphy replied:
That seems dissonant though. In the README, we outline that true or false are valid options for this attribute, but here we fail with a validation error if false is actually used.

alecxvs replied:
b44cde1ad7ecc81a3bd9c73085de042e587905ed

jdmurphy commented on test/unit/spec/default_spec.rb:
```diff
@@ -0,0 +1,305 @@
+              )
+            end
+
+            it { is_expected.to create_remote_file(package_dir).with(source: urls[:splunk]) }
```
We are mixing and matching test formats all over which irks me; it seems to me that it would be cleaner, more readable and easier to update in the future if we stuck with one format. I personally think we should stick with the following format:
```
it 'should do something' do
  expect(something).to happen
end
```

jdmurphy replied:
Fine...personal preference. You like that style, I like the other. If we can't do all one style though, then it is the mix and match which I still think we need to fix.

alecxvs replied:
Unfortunately it's impossible to use one-liners throughout, so we'll have to use the standard format.

alecxvs replied:
c75a9507b2767e7ce0df595009bcfe7de35ffcb4

jdmurphy commented on test/unit/spec/default_spec.rb:
```diff
@@ -0,0 +1,305 @@
+
+            it { is_expected.to create_remote_file(package_dir).with(source: urls[:splunk]) }
+
+            case platform
```
I can see this kind of block multiple places in the test file. We should be able to move this to a shared_example to help clean and DRY this file up some.

alecxvs replied:
i'm not sure how I would accomplish that. There are multiple case platform blocks, but each of them have different conditions, and I don't know how I would combine all of them and still reduce code.

jdmurphy replied:
At the very least you can have a shared example for installation of packages that relies on platform, and any parameters used in the install block. I can see that kind of thing all over this file. In addition, you can have shared contexts for things such as your contexts around "with the Splunk package" and "with the Universal Forwarder package". There are lots of ways we can clean this up.

alecxvs replied:
I'm working on combining the package contexts.

alecxvs replied:
c75a9507b2767e7ce0df595009bcfe7de35ffcb4

jdmurphy commented on test/unit/spec/default_spec.rb:
```diff
@@ -0,0 +1,305 @@
+            end
+
+            context 'when Splunk is already installed' do
+              let(:do_before) do
```
Whats the reasoning for using this instead of the regular 
```
before do
...
end
```

blocks?

alecxvs replied:
A before block would run before each test, but oddly not before the subject (the chef run). I need to be able to stub things before the chef run too, so this let is called before the chef run is converged.

gravesb replied:
I'm not sure I follow your reasoning.  All the tests still pass if I switch this to a regular before do block.  Unless you accidentally broke something along the way during the review, the chef_run is currently being executed for every example, not just once per example group.

alecxvs replied:
I need to look into that, because chef runs should not be per example.

alecxvs commented:
Squashing commits ⚡

acharlieh commented on resources/splunk_install.rb:
```diff
@@ -0,0 +1,215 @@
+  end
+
+  action :uninstall do
+    super()
```
We should have super() be the last call in each uninstall... remove package then clean up directory leftovers.

alecxvs replied:
3685b1a00572aee2694b768382963aae1f1999df

acharlieh commented on README.md:
```diff
@@ -0,0 +1,69 @@
+| version        |               String                |       **Yes**        |                                          | Version of Splunk to install                                                                                                                                                                                                                                           |
+| build          |               String                |       **Yes**        |                                          | Build number of the version                                                                                                                                                                                                                                            |
+| base_url       |               String                |          No          | `'https://download.splunk.com/products'` | Base url to pull Splunk packages from. Use this if you are mirroring the downloads for Splunk packages. The resource will append the version, os, and filename to the url like so: `{base_url}/splunk/releases/0.0.0/linux/splunk-0.0.0-a1b2c3d4e5f6-Linux-x86_64.tgz` |
+| accept_license |                true                 | **Yes (on Windows)** |                                          | Set if you accept Splunk's license agreement on install. Currently only for the Windows Package, as it requires the license to be accepted at install time. On Splunk for Linux, it can be accepted on first run with `--accept-license`.                              |
```
I've been thinking about this a lot... It's a parameter, that's required to be set to true in a package designed for automation. What if we removed this parameter, and instead put something to the effect of "Using resources of this cookbook to install and run Splunk means you agree to Splunk's EULA packaged with the software, online at http://www.splunk.com/en_us/legal/splunk-software-license-agreement.html" Then we can get rid of the option, and the logic reacting to having different options for it.

jdmurphy replied:
+1

alecxvs replied:
3312faca73163e7382aef00a074d78cf5e16943a

acharlieh commented on README.md:
```diff
@@ -0,0 +1,69 @@
+##### Action *:uninstall*
+Removes Splunk or Universal Forwarder and all its configuration.
+
+Properties: _None_
```
Realized this isn't quite right... you need the package name here.

alecxvs replied:
18fa620bef53ba127a1daa4e97f074a16ee451b1

acharlieh commented:
Few minor changes left, then 👍 from me.

acharlieh commented on resources/splunk_install.rb:
```diff
@@ -0,0 +1,215 @@
+  default_action :install
+
+  def after_created
+    node.run_state['splunk_ingredient'] ||= { 'installations' => {} }
```
we should separate the initialization of splunk_ingredient, and splunk_ingredient.installations run states... the thought being with future resources, what if someone uses a different resource that doesn't deal with installations first?

alecxvs replied:
I wonder if that change would be better made later when there is actually another resource using the run state. It would be trivial to extract this into a helper method and remain passive later, and I can't determine the design for very future resources yet. It could be the case that the run state for every other resource exists under the context of an installation (because logically there might not be anything we're managing outside of installed Splunk).

acharlieh replied:
Good call... I was definitely over thinking that one.

gravesb commented on README.md:
```diff
@@ -0,0 +1,76 @@
+
+| Name    |               Type(s)               | Required | Default | Description                                                                                                                                                                                       |
+|:--------|:-----------------------------------:|:--------:|:--------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| package | `:splunk` or `:universal_forwarder` | **Yes**  |         | Specifies the Splunk package to install. You must specify the package, or name the resource for the package; for example, `package :splunk` or `splunk_install 'universal_forwarder' do ... end`. |
```
minor, "...the splunk package to uninstall"

alecxvs replied:
146f45c

gravesb commented on resources/splunk_install.rb:
```diff
@@ -0,0 +1,214 @@
+
+  action :install do
+    install_state = node.run_state['splunk_ingredient']
+    raise "Install at #{install_dir} already exists!" if install_state['installations'][install_dir]
```
So would this only happen if a node's run list contained this install resource multiple times right? Or I suppose if the install action was notified multiple times during a run.

alecxvs replied:
Yes, executing the install action multiple times pointing to the same installation path (currently, basically the same package) will throw this error. You can run any other actions (like uninstall) without encountering this error, but attempting to install it twice will be an error.

It should be possible to run the install, uninstall, then install in the same chef run without encountering this error, but that's weird and isn't tested.

gravesb commented on resources/splunk_install.rb:
```diff
@@ -0,0 +1,214 @@
+      windows_package package_name do
+        source package_path.to_s
+        action :install
+        options 'LAUNCHSPLUNK=0 INSTALL_SHORTCUT=0 AGREETOLICENSE=Yes'
```
So from the doc, install_shortcut only applies for the full splunk installation.  I'm not sure if it hurts anything to pass in additional flags but we could pretty easily flex the options based on the package attribute.

alecxvs replied:
Integration tests succeed so it doesn't seem to adversely affect the installation. If we add more options specific to the packages in the future, we should flex it, but that's not really necessary here.

gravesb commented:
Since we only support ruby 2.x (chef 12.x +) you can drop all the encoding comments in the source files.  utf-8 is the default in ruby 2+.

gravesb commented on .kitchen.yml:
```diff
@@ -0,0 +1,32 @@
+    run_list:
+      - recipe[cerner_splunk_ingredient_test::splunk]
+      - recipe[cerner_splunk_ingredient_test::universal_forwarder]
+  - name: uninstall
```
We should have a third test kitchen suite to test uninstalling the universal forwarder.  We should rename this suite (and the uninstall recipe) to be more specific to the splunk package.  e.g. uninstall_splunk

alecxvs replied:
49878a781929b6afc2aaa898a1b457fb6b50a11f

I added a recipe for universal forwarder and added it to the uninstall suite; is there any reason to split the uninstall suite for the two packages? There is no difference test-wise, and I intended to cut down on test time by not setting up chef omnibus on ~20 VMs.

gravesb replied:
No, that's a good call.  I just didn't think about adding the recipes to the same suite.

alecxvs commented:
I did not know that 👍

gravesb commented on test/unit/spec/default_spec.rb:
```diff
@@ -0,0 +1,121 @@
+
+    context 'when package is not specified' do
+      let(:test_params) { { name: 'hotcakes', action: :uninstall } }
+      it 'fails to install' do
```
nit, fails to uninstall.  Or possibly reword to just 'fails the chef run' in which case you could do the same for all the other failure tests so they can all be safely copy pasted around.

alecxvs replied:
2a95c28a6d482f92890db0aab7f5792c5210018d

gravesb commented on resources/splunk_install.rb:
```diff
@@ -0,0 +1,214 @@
+
+  def package_name
+    @package_name ||= package_names[package][node['os'].to_sym]
+    raise "Unsupported Combination: #{package} + #{node['os']}" unless @package_name
```
need chef specs for this raise error. same for the raise in install_dir below.

alecxvs replied:
I can test for the error, but I don't think I can test these methods specifically.

alecxvs replied:
521ee580f823588b64ce4b075f5448d9da3b8ed0

gravesb replied:
Yeah, that's what I meant.  A test that causes this line to be executed.

gravesb replied:
Follow up question, which one of these raise statements does that test hit?  I think this one is dead code since the install_dir method is currently always called before this one (if I'm manually tracing the code paths correctly).

alecxvs replied:
Yeah, that will be called first in any case.

gravesb commented on test/unit/spec/shared_examples.rb:
```diff
@@ -0,0 +1,127 @@
+      end
+    end
+
+    context "when '#{package}' is already installed" do
```
can you add a context for when the package is already installed but with a different version?

alecxvs replied:
038f2995d0210267348539f99f18742c49cdaf9f

alecxvs commented:
9d56b606c536f7b5c80dd398b369360ceea5d6b7

Removed encoding comments

jdmurphy commented on test/integration/uninstall/spec.rb:
```diff
@@ -0,0 +1,27 @@
+  end
+
+  describe command('ps -eo comm= | grep splunk') do
+    its('exit_status') { should eq 1 }
```
Just wondering, why the difference between install and uninstall for `is_expected.to eq 1` and `should eq 1`?

alecxvs replied:
Copy paste error, probably

alecxvs replied:
6d76bbef842b7e565fc79c86e180b5f359bd8a94

jdmurphy commented on .travis.yml:
```diff
@@ -0,0 +1,14 @@
+language: ruby
+cache: bundler
+sudo: false
+rvm:
```
Do we really need separate testing for these three versions of ruby? At most I could see 2.2.x vs 2.3.x, but even those should be remarkably similar.

gravesb replied:
That's a good point, I think chef's embedded ruby is still 2.1.x.  I have chef-dk 0.13.21 that comes packaged with chef 12.9.41 and ruby 2.1.8. So I think this should definitely include ruby 2.1.<something>.

alecxvs replied:
Really? I expected they would be using a newer Ruby by now.
Regardless, Guard in the bundle requires >= 2.2.3 and since travis installs the whole gem bundle, I can't test 2.1.x. I will limit it to just 2.2.5.

4328e2192592966f6e3a692e192431207df046fc

acharlieh replied:
I had missed your reply [to my previous comment](https://github.com/cerner/cerner_splunk_ingredient/pull/1#discussion_r65640587) but we should indeed be running tests under ruby 2.1.x lest we have some syntax or quirk that doesn't jive with the version of ruby that ships with Chef. 

Guard seems like a utility that you'd use only with local development... should it be in the bundle? Or should we have a way to exclude it for Travis builds? Also there's an interesting note on https://github.com/guard/guard saying that with the right version of bundler, it should install the appropriate version for your ruby... maybe that's something to look into as well?

jdmurphy replied:
Couldn't we just move guard and all the related stuff to a development block and do bundle install without development?

alecxvs replied:
Is that possible with travis? I don't specify the bundle install command, just that it should use bundler.

jdmurphy replied:
See [this page](https://docs.travis-ci.com/user/languages/ruby). It describes a `bundler_args` option which sounds like exactly what we need.

alecxvs replied:
Exactly what we need indeed. I can cut out kitchen gems too since we don't use those in travis.

I had no success installing guard on 2.1.8 or 2.1.10. They say bundler should be able to resolve it, but it doesn't. So specifying the args is necessary unless I remove the guard gems from the bundle entirely (seems weird to me given the Guardfile referencing Guard gems for RSpec and Foodcritic)

jdmurphy replied:
Rockets haven't been necessary (barring having some characters in the key) since Ruby 2.0 I believe.

alecxvs replied:
I do have special characters in the key though, "7.2" for example. "7.2": "value" translates to a symbol key (and there is no way to specify a string key without a rocket, but I need to rewrite a lot of stuff to expect strings instead of symbols now anyways)

alecxvs replied:
b2f678ca0339c5f41bdefa387dafa2d64d320d3a

alecxvs commented:
Referencing this comment out here because it's buried in an outdated diff on a squashed commit

https://github.com/cerner/cerner_splunk_ingredient/pull/1#discussion_r66867234

> I'm not sure I follow your reasoning. All the tests still pass if I switch this to a regular before do block. Unless you accidentally broke something along the way during the review, the chef_run is currently being executed for every example, not just once per example group.

The main point is that apparently chef_run is converging for every example (because chef_run.converge is specified in a let). I'm not sure why before was not working for me previously, but it shouldn't be working for me now (and it is). Converging for every example is an absolute waste since ChefSpec only checks that resources were called and each example does not need to be isolated from each other. Furthermore, I wrote the contexts to separate differences in state between all the test cases.

I'm trying to devise a way of defining the chef_run for each example group, but there's no broader let. The closest thing I have is a subject, but I will have to copy paste it for every context and that doesn't solve my wasted computation time with the dozen other lets that are only specified and used once per context.
An effective solution would be collapsing all the examples in my context under one it block, but that makes for a terrible test output.

alecxvs commented:
I'm going to switch most of the contexts so that expectations under a single chef run all reside under the same example. The output won't be great, but it will make more sense as far as determining when chef runs happen during unit tests, and it will speed up unit tests tremendously.

The alternative, being to only converge in example groups, ends up being terribly complex.

gravesb commented:
Why PR into a stable branch instead of just using master (and making master only contain stable code)?

gravesb commented on README.md:
```diff
@@ -0,0 +1,76 @@
+Requirements
+------------
+Chef >= 12.4
+Ruby >= 2.2.3
```
need to adjust this after our ruby version discussion earlier on this PR. Since chef 12.4.3 packages ruby 2.1.<something> we can't say we require 2.2.3 or above.

alecxvs replied:
1f9d5e96d28047ec711c78d7f9a1f74a9a9e55b3

acharlieh commented:
@gravesb With regards to the `stable` versus `master` I figured we could start with following the same pattern as [cerner_splunk](https://github.com/cerner/cerner_splunk)... that said there could be an argument to be made to switch branches about in both repos.

alecxvs commented:
I don't see the practical purpose of having stable and master with master sitting at a tagged version. The tagged versions are suitable for getting releases, and if we really want to have a branch consistently pointing to the latest release, a branch named `latest` would be more concise.

acharlieh commented:
@alecxvs it came from trying to prevent people from accidentally taking unreleased code when pointing at the github repo (prior to when we were using Supermarket internally to upload cookbooks to chef servers) while we tried incremental changes in process for contributing to cerner_splunk. If you're interested in the history, check out cerner/cerner_splunk#41.

gravesb commented:
Not a big deal, just personal preference to follow other open source cookbooks (and the vast majority of git projects in general) that uses master for the latest reviewed/approved code.  Per Charlie's comment, given the supermarket, I also don't see a need for a branch that always points to the latest release.

+1

alecxvs commented:
If there are no objections, I would prefer merging into master and getting rid of stable at the conclusion of this pull request.

alecxvs commented:
I think you're right about waiting, we can rearrange the branches after closing this.

alecxvs committed:
Initial Project Structure

alecxvs committed:
Implement splunk_install resource

alecxvs committed:
Unit Tests

alecxvs committed:
Integration Tests

jdmurphy commented:
+1 then if we are just going to wait and rearrange later.

alecxvs commented:
Squashed 💥
